docs/design/master0-recovery.md

```diff
@@ -0,0 +1,111 @@
+
+1. expose below `Registry` struct to clusterfile which is contains configurations about local registry and remote
+   registry.
+
+```yaml
+type Registry struct {
+  Domain        string        `json:"domain,omitempty"`
+  Port          int           `json:"port,omitempty"`
+  Username      string        `json:"username,omitempty"`
+  Password      string        `json:"password,omitempty"`
+  LocalRegistry LocalRegistry `json:"local_registry,omitempty"`
+}
+  type LocalRegistry struct {
+  // DeployHosts is the target host list that local registry will be deployed on.
+  // if not set ,master0 will be the default value.
+  DeployHosts []net.IP `json:"deploy_hosts,omitempty"`
+  // InsecureMode indicated that whether the local registry is exposed in HTTPS.
+  // if true sealer will generate default ssl cert.
+  InsecureMode bool `json:"insecure_mode,omitempty,omitempty"`
+}
+```
+
+2. save registry configuration to cluster as configmap with name "sealer-registry" in namespace "kub-system".
+
+```shell
+# must exist, this is core registry configuration
+/var/lib/sealer/data/my-cluster/rootfs/etc/registry.yml
+#if local registry is not booted on insecure mode
+/var/lib/sealer/data/my-cluster/rootfs/certs/sea.hub.crt
+/var/lib/sealer/data/my-cluster/rootfs/certs/sea.hub.key
+# if local registry is booted with basic auth.
+/var/lib/sealer/data/my-cluster/rootfs/etc/registry_htpasswd
+```
+
+3. save boot history of application image to cluster as configmap with name "sealer-apps" in namespace "kub-system".
+
+```json
+{
+  "type": "kube-installer",
+  "applications": [
+    {
+      "name": "nginx",
+      "image_id": "sha256:6bc6bc3015103b5a20d25f80cc833c0df5c5fccfdcecacfedcea296c616a534e",
+      "type": "helm",
+      "launch_cmd": "helm install nginx application/apps/nginx/",
+      "version": "v1"
+    },
+    {
+      "name": "dashboard",
+      "image_id": "sha256:6bc6bc3015103b5a20d25f80cc833c0df5c5fccfdcecacfedcea296c616a534e",
+      "type": "kube",
+      "launch_cmd": "kubectl apply -f application/apps/dashboard/",
+      "version": "v1"
+    }
+  ],
+  "launch": {
+    "app_names": [
+      "nginx",
+      "dashboard"
+    ]
+  }
+}
+```
+
+### For recovery of builtin registry
+
+1. Pre-distribute all registry data to all deploy hosts
+
+* load registry configuration from cluster on new master0
+* dump registry configuration files from cluster to rootfs on new master0
+* launch builtin registry thought rootfs scripts and configuration files to recover registry service.
+* replace the latest registry ip address on all cluster nodes.
+
+2. Post-distribute registry data on recovery host
+
+* load registry configuration from cluster on new master0
+* dump registry configuration files from cluster to rootfs on new master0
+* distribute all registry data to recovery host thought boot history loaded from cluster on new master0
+* launch builtin registry thought rootfs scripts and configuration files to recover registry service.
```

Comment:
I prefer to distribute registry data to each master, so that each master is equal. If use deployhost, do users have to manually handle the distributing the registry to another new deployhost after recovering once?

Reply:
i agree with the point that make all masters is consistent,  `post-distribute` is more flexibility than `pre-distribute` , Both needs to manually  specify on which node the registry service is recover to. The difference is that `pre-distribute` can save the original same data with master0, and the `post-distribute` save the later recovered data according to the history.

Reply:
we can implemet pre-distribute firstly.